tarantool_test.go

```diff
@@ -342,6 +373,131 @@ func BenchmarkClientParallel(b *testing.B) {
+	defer conn.Close()
+
+	_, err := conn.Replace(spaceNo, []interface{}{uint(1111), "hello", "world"})
+	if err != nil {
+		b.Fatal("No connection available")
+	}
+
+	req := NewSelectRequest(spaceNo).
+		Index(indexNo).
+		Limit(1).
+		Iterator(IterEq).
+		Key([]interface{}{uint(1111)})
+
+	b.SetParallelism(multiplier)
+	b.ResetTimer()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			_ = conn.Do(req)
+			_, err := conn.Do(req).Get()
+			if err != nil {
+				b.Error(err)
+			}
+		}
+	})
+}
+
+func benchmarkClientParallelRequestObjectWithContext(multiplier int, b *testing.B) {
+	conn := test_helpers.ConnectWithValidation(b, server, opts)
+	defer conn.Close()
+
+	_, err := conn.Replace(spaceNo, []interface{}{uint(1111), "hello", "world"})
+	if err != nil {
+		b.Fatal("No connection available")
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	req := NewSelectRequest(spaceNo).
+		Index(indexNo).
+		Limit(1).
+		Iterator(IterEq).
+		Key([]interface{}{uint(1111)}).
+		Context(ctx)
+
+	b.SetParallelism(multiplier)
+	b.ResetTimer()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			_ = conn.Do(req)
+			_, err := conn.Do(req).Get()
+			if err != nil {
+				b.Error(err)
+			}
+		}
+	})
+}
+
+func benchmarkClientParallelRequestObjectMixed(multiplier int, b *testing.B) {
+	conn := test_helpers.ConnectWithValidation(b, server, opts)
+	defer conn.Close()
+
+	_, err := conn.Replace(spaceNo, []interface{}{uint(1111), "hello", "world"})
+	if err != nil {
+		b.Fatal("No connection available")
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	req := NewSelectRequest(spaceNo).
+		Index(indexNo).
+		Limit(1).
+		Iterator(IterEq).
+		Key([]interface{}{uint(1111)})
+
+	reqWithCtx := NewSelectRequest(spaceNo).
+		Index(indexNo).
+		Limit(1).
+		Iterator(IterEq).
+		Key([]interface{}{uint(1111)}).
+		Context(ctx)
+
+	b.SetParallelism(multiplier)
+	b.ResetTimer()
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			_ = conn.Do(req)
+			_, err := conn.Do(reqWithCtx).Get()
+			if err != nil {
+				b.Error(err)
+			}
+		}
+	})
+}
+
+func BenchmarkClientParallelRequestObject(b *testing.B) {
+	multipliers := []int{10, 50, 500, 1000}
+	conn := test_helpers.ConnectWithValidation(b, server, opts)
+	defer conn.Close()
+
+	_, err := conn.Replace(spaceNo, []interface{}{uint(1111), "hello", "world"})
+	if err != nil {
+		b.Fatal("No connection available")
+	}
+
+	for _, m := range multipliers {
+		goroutinesNum := runtime.GOMAXPROCS(0) * m
+
+		b.Run(fmt.Sprintf("Plain        %d goroutines", goroutinesNum), func(b *testing.B) {
+			benchmarkClientParallelRequestObject(m, b)
+		})
+
+		b.Run(fmt.Sprintf("With Context %d goroutines", goroutinesNum), func(b *testing.B) {
+			benchmarkClientParallelRequestObjectWithContext(m, b)
+		})
+
+		b.Run(fmt.Sprintf("Mixed        %d goroutines", goroutinesNum), func(b *testing.B) {
+			benchmarkClientParallelRequestObjectMixed(m, b)
```

Comment:
It would not be fair, that is why each of mentioned benchmarks has one asynchronous `conn.Do()` and one synchronous `conn.Do().Get()` call. That is needed to mix futures with and without context in the internal queue in case of ObjectMixed. For Object/ObjectWithContext it is needed to have the same load of Futures as in ObjectMixed and for fair comparing the results.

Reply:
How exactly do we compare the results?

Reply:
After running benchmarks and process it with benchstat tool, you may get such [results](https://github.com/tarantool/go-tarantool/pull/173#issuecomment-1186955624). This benchmark is intended to show if using contexts in requests leads to perf-degradation. You can watch the difference between using only requests with nil context, with not nil context and the mixed usage.

Reply:
So why then ClientParallelRequestObject/Plain________40_goroutines, ClientParallelRequestObject/With_Context_40_goroutines, ClientParallelRequestObject/Mixed________40_goroutines time/op is similar and not Mixed=Plain+With_Context?

Reply:
```
_ = conn.Do(req)
_, err := conn.Do(reqWithCtx).Get()
```

So we doesn't wait for the first one to finish?